src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/__Finalizer.cs

```diff
@@ -63,10 +64,15 @@ private static unsafe uint DrainQueue()
 
                 finalizerCount++;
 
-                // Call the finalizer on the current target object. If the finalizer throws we'll fail
-                // fast via normal Redhawk exception semantics (since we don't attempt to catch
-                // anything).
-                ((delegate*<object, void>)target.GetMethodTable()->FinalizerCode)(target);
+                try
+                {
+                    // Call the finalizer on the current target object.
+                    ((delegate*<object, void>)target.GetMethodTable()->FinalizerCode)(target);
+                }
+                catch (Exception ex) when (ExceptionHandling.s_handler != null && ExceptionHandling.s_handler(ex))
+                {
+                    // the handler returned "true" means the exception is now "handled" and we should continue.
+                }
```

Comment:
If we ever add a mode of the runtime where multiple native AOT compiled native libraries share the GC, would we have multiple finalizer threads for each library, or just one finalizer thread shared by all? For example, the Redhawk design was a single finalizer thread.

In a single finalizer thread design this wouldn't work since there can only be one handler that would be getting exceptions from all System.Object universes.

We do have indirection facilities to do callbacks into the appropriate classlib in the Redhawk design - look for `RhpGetClasslibFunctionFromEEType` and `ClasslibFunctionId` enum. It should be a mechanical change to switch to that approach provided that's the design we want to keep.

Reply:
Hmm, actually, the whole try/catch would need to move behind a `RhpGetClasslibFunctionFromEEType` callback.

The `catch (Exception ex)` would only catch exceptions from the type universe that defines the finalizer thread.

It doesn't look like it's possible to express the kind of filter we need in C#. We would want:

```
catch when (ExceptionHandling.s_handler != null && ExceptionHandling.s_handler(value)) // does not compile because value is undeclared
```

where `value` is the thing that came into the filter. C# compiler will currently generate an annoying isinst at the beginning of the filter block, checking for `System.Exception`.

Reply:
Setting Unhandled Exception Handler is supposed to be a hosting-like API with process-wide effects. 

We will have to think about a general model of how process-wide features work (or not?) in the context of multiple individually-trimmed things (domains?) that share the same runtime/GC/finalizer, while cannot share each other`s types/objects/threads/locks/...

I did think about this case, but so far I do not have anything better than just not supporting the feature (as in `SetUnhandledExceptionHandler` throws something like `NotSupportedInMulimoduleMode`). That would be easy to do, but perhaps a better solution will be found when we actually get to supporting such mode.

Reply:
This hander should not work like a process-wide feature. It should be only applicable to the managed world that it is part of. It should not be handling exceptions in foreign managed worlds.

In any case, it is not something that needs to solved as part of this PR.